### PR TITLE
Facilitator interfaces and basic repo setup

### LINKED ISSUE #5: Define the interfaces for all the functions in Facilitator
Worker facilitates the stake and mint process. 
- It reads the stake requested data from the database.
- It gets the worker address for the database for each chain.
- It unlocks the worker address account.
- It updates the database with the current status of facilitation.
For more details check the epic #1

# **This ticket is about defining the interfaces of all the functions in the Worker.**

The proposed functions are provided for reference.
The implementer can think this in detail and can add/remove functions.

- [ ] **Constructor**
	<strike>- Accepts `mosaic-config`, `facilitator-config` as parameter.</strike>
	- Accepts Config object. (store in variable).
	- Creates DB object. (store in variable). (Still in discussion)
- [ ] **function to start the worker**.
<strike>
- [ ] **function to get pending items from `stake_request` table. (Create the ticket for this.)
- [ ] **function that handles the state machine** (Create a new ticket for this.)
- [ ] **function to get the message status from messageOutbox of the gateway.**
- [ ] **function to get the message status from messageInbox of the cogateway.**
- [ ] **function to get the latest `stateRoot` from the anchor contract (auxiliary chain).**

- [ ] **function to update DB with the latest state of facilitation.** (Task execution table) (Update the ticket).
	<strike>- REQUESTED 
	- ACCEPTED 
	- ACCOUNT_PROVEN 
	- STAKE_INTENT_COMFIRMED 
	- PROGRESS_STAKE 
	- PROGRESS_MINT 
	- COMPLETE </strike>

- [ ] **function to update DB with `hashlock` and `unlockSecret`.**
<strike>- [ ] **function to update DB with proof data.**</strike>

<strike>
- [ ] **function to unlock worker address.**
	- Get the encrypted_account from Config for the given address.
	- Read the password from the env variable.
	- Decrypt the account.
	- Add it to web3 account wallet.
</strike>
The accounts will be unlocked at the commander level and will update the web3 wallet.

- [ ] **function to create `hashLock` and `unlockSecret`.**
	- Generate hashlock and unlockSecret pair

- [ ] **function to approve bounty amount.**
- [ ] **function to accept stake request.**
- [ ] **function to generate account proof.**
- [ ] **function to confirm stake intent.**
- [ ] **function to progress stake.**
- [ ] **function to progress mint.**

</strike>
- [ ] **function to stop the worker (<strike>handle SIGTERM</strike>).** (update the ticket.)

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below:

- Facilitator class and function declarations.

- Config class as placeholder

- Basic repo setup files:
  - travis.yml
  - eslint config
  - tsconfig.json
  - package.json
  - package-lock.json
  - .gitignore file

Fixes #5

### CODE REVIEW

File: src/Facilitator.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import Config from './Config';
+
+/**
+ * The class defines properties and behaviour of facilitator.
```
Comment (sarvesh-ost) [inline src/Facilitator.ts]:
```suggestion
 * The class defines properties and behaviors of a facilitator.
```

File: src/Facilitator.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  private dbConnection;
+
+  /**
+   * Facilitator class constructor.
+   *
+   * @param {string} config Config class object.
+   * @param {object} dbConnection DB connection object.
+   * @param {string} chainId Chain identifier to subscribe.
+   */
+  public constructor(config, dbConnection, chainId) {
```
Comment (sarvesh-ost) [inline src/Facilitator.ts]: What are the types of `config`, `dbConnection` and `chainId` ?

File: src/Facilitator.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * The class defines properties and behaviour of facilitator.
+ */
+export default class Facilitator {
+  private config: Config;
+
+  private chainId: string;
+
+  private dbConnection;
```
Comment (sarvesh-ost) [inline src/Facilitator.ts]: What's the data type?

File: src/DBConnection.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/**
+ * The class creates db connection object using npm client.
+ */
+export default class DBConnection {
+  private static connection;
```
Comment (sarvesh-ost) [inline src/DBConnection.ts]: Data type?

File: src/DBConnection.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  /**
+   * Creates and returns db connection object.
+   */
+  public static getConnection(config) {
```
Comment (sarvesh-ost) [inline src/DBConnection.ts]: Please add parameter and return type

Comment (sarvesh-ost) [reply src/DBConnection.ts]: Maybe this doesn't need the whole config. Just connection string, user and password should be sufficient.

File: src/Config.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/**
+ * The class encapsulates properties and behaviour of mosaicConfig and facilitatorConfig.
+ */
+export default class Config {
+  /**
+   * Config class constructor.
+   */
+  public constructor() {
```
Comment (sarvesh-ost) [inline src/Config.ts]: Do we need a public constructor in the config?   There should a factory or something which should instantiate config possibly from file.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    "Token Economy",
+    "web3",
+    "Ethereum",
+    "EIP20"
+  ],
+  "homepage": "https://openst.org",
+  "author": "OpenST Ltd.",
+  "license": "LGPL-3.0-only",
+  "bugs": {
+    "url": "https://github.com/openst/mosaic.js/issues"
```
Comment (sarvesh-ost) [inline package.json]: Please correct URL

File: package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  "scripts": {
+    "lint": "eslint src -c .eslintrc.json --ext ts"
+  },
+  "dependencies": {
+    "typescript": "^3.3.3333"
```
Comment (sarvesh-ost) [inline package.json]:
```suggestion
    "typescript": "3.3.3333"
```
Is it a dev dependency?

File: package.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  "devDependencies": {
+    "@typescript-eslint/eslint-plugin": "^1.5.0",
```
Comment (sarvesh-ost) [inline package.json]:
```suggestion
    "@typescript-eslint/eslint-plugin": "1.5.0",
```
Removed `^` from all external dependencies.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    "Mosaic",
+    "Simple Token",
+    "Token Economy",
+    "web3",
+    "Ethereum",
+    "EIP20"
+  ],
+  "homepage": "https://openst.org",
+  "author": "OpenST Ltd.",
+  "license": "LGPL-3.0-only",
```
Comment (benjaminbollen) [inline package.json]: Apache v2.0

File: src/Facilitator.ts (hunk 2 above)
Comment (benjaminbollen) [inline src/Facilitator.ts]: is the chainId needed if the facilitator manifest already specifies the chain(s) to facilitate?

File: src/Config.ts (hunk 6 above)
Comment (benjaminbollen) [reply src/Config.ts]: this overlaps also with #63 ?

**Comment by abhayks1** [comment]
Started addressing the feedback 😼

File: src/Config.ts (hunk 6 above)
Comment (abhayks1) [reply src/Config.ts]: PR #63 is merged. Rebased my branch with develop.

File: package.json (hunk 8 above)
Comment (abhayks1) [reply package.json]: Not 100% sure. In mosaic chains also, it's in dependency.

File: src/Facilitator.ts (hunk 2 above)
Comment (abhayks1) [reply src/Facilitator.ts]: makes sense 💯 . 
I saw in the ticket that while starting facilitator chainID is specified on command line.
https://github.com/OpenST/facilitator/issues/6
I think that's not needed. Will make the changes.

**Comment by abhayks1** [comment]
Available for review again 🎸

**Comment by deepesh-kn** [comment]
Reviewing: 🦖

**Pull request merged**